crates/flipperzero/src/furi/sync.rs

```diff
@@ -1,82 +1,108 @@
+    unsafe fn create(&self) -> *mut sys::FuriMutex {
+        let mutex = unsafe { sys::furi_mutex_alloc(MUTEX_TYPE) };
+
+        match self
+            .0
+            .compare_exchange(ptr::null_mut(), mutex, Ordering::Release, Ordering::Relaxed)
+        {
+            Ok(_) => mutex,
+            Err(global_ptr) => {
+                unsafe { sys::furi_mutex_free(mutex) };
+                global_ptr
+            }
         }
+    }
 
-        Ok(MutexGuard(self, PhantomData))
+    /// Attempts to acquire the mutex within `timeout` ticks, or polls without acquiring
+    /// if `timeout` is zero.
+    fn try_acquire(&self, timeout: u32) -> bool {
+        let status: Status = unsafe { sys::furi_mutex_acquire(self.get(), timeout).into() };
+        status.is_ok()
     }
 }
 
-unsafe impl<T: ?Sized + Send> Send for Mutex<T> {}
-unsafe impl<T: ?Sized + Send> Sync for Mutex<T> {}
-
-/// An RAII implementation of a "scoped lock" of a mutex.
-/// When this structure is dropped (falls out of scope), the lock will be unlocked.
-pub struct MutexGuard<'a, T: ?Sized + 'a>(&'a Mutex<T>, UnsendUnsync);
+impl Drop for FuriMutex {
+    fn drop(&mut self) {
+        let mutex = self.0.load(Ordering::Acquire);
+        if !mutex.is_null() {
+            unsafe { sys::furi_mutex_free(mutex) };
+        }
+    }
+}
 
-impl<T> Deref for MutexGuard<'_, T> {
-    type Target = T;
+unsafe impl RawMutex for FuriMutex {
+    const INIT: Self = FuriMutex::new();
+    type GuardMarker = GuardNoSend;
 
-    fn deref(&self) -> &Self::Target {
-        unsafe { &*self.0.data.get() }
+    fn lock(&self) {
+        // `INCLUDE_vTaskSuspend` is set to 1 in the Flipper Zero's FreeRTOS config, so as
+        // long as this timeout value matches `portMAX_DELAY` in the FreeRTOS config, this
+        // will block indefinitely as intended.
+        assert!(self.try_acquire(u32::MAX));
     }
-}
 
-impl<T> DerefMut for MutexGuard<'_, T> {
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        unsafe { &mut *self.0.data.get() }
+    fn try_lock(&self) -> bool {
+        // Poll first so we don't block at all if we cannot acquire.
+        self.try_acquire(0) && self.try_acquire(1)
     }
-}
 
-impl<T: ?Sized> Drop for MutexGuard<'_, T> {
-    fn drop(&mut self) {
-        let status: Status = unsafe { sys::furi_mutex_release(self.0.mutex.as_ptr()).into() };
+    unsafe fn unlock(&self) {
+        let status: Status = unsafe { sys::furi_mutex_release(self.get()).into() };
         if status.is_err() {
             panic!("furi_mutex_release failed: {}", status);
         }
     }
 }
 
-// `UnsendUnsync` is actually a bit too strong.
-// As long as `T` implements `Sync`, it's fine to access it from another thread.
-unsafe impl<T: ?Sized + Sync> Sync for MutexGuard<'_, T> {}
+unsafe impl RawMutexTimed for FuriMutex {
+    type Duration = u32; // TODO replace with `Duration` type that hides ticks.
+    type Instant = u32; // TODO replace with `Instant` type that hides ticks.
```

Comment:
I opened this PR now so it can be reviewed, but we probably don't want to merge it before we have decided on a solution for these types.

Reply:
In the short term, you could create a `Duration(u32)` and `Instant(u32)` new-type so that the implementation is opaque.

This can then be replaced with real `Duration` and `Instant` types once we are happy with an approach.